### ISSUE TITLE
Fallback when SetThreadDescription isn't available.

### ISSUE THREAD

**Issue by mtwilliams** [BuckeyeSoftware/rex#1]
`SetThreadDescription` wasn't introduced until Windows 10, 1607. It's [currently being used](https://github.com/BuckeyeSoftware/rex/blob/cbc8e68ae0910530a85d65f7a13cc3b897521331/src/rx/core/concurrency/thread.cpp#L84) to name threads without a check to determine if it's actually available. Instead, it should be fetched dynamically with a fallback to the older approach that only worked under debuggers.

I suggest a static initializer that does something akin to:

```cpp
SetThreadDescription set_thread_description;
*(void **)&set_thread_description =
  GetProcAddress(GetModuleHandleA("kernel32.dll"), "SetThreadDescription");
if (!set_thread_description)
  set_thread_description = &set_thread_description_fallback;
```

---

Here's the fallback, for Windows 7.

```cpp
if (IsDebuggerPresent()) {
  #pragma pack(push, 8)
    typedef struct THREAD_NAME_INFO {
      DWORD dwType;
      LPCSTR szName;
      DWORD dwThreadID;
      DWORD dwFlags;
    } THREAD_NAME_INFO;
  #pragma pack(pop)

  THREAD_NAME_INFO thread_name_info;

  thread_name_info.dwType = 0x1000;
  thread_name_info.szName = name;
  thread_name_info.dwThreadID = GetCurrentThreadId();
  thread_name_info.dwFlags = 0;

  #pragma warning(push)
  #pragma warning(disable: 6320 6322)
    __try {
      // This is how you used to name threads in Windows. Talk about a mess.
      RaiseException(0x406D1388, 0, sizeof(THREAD_NAME_INFO)/sizeof(ULONG_PTR), (ULONG_PTR *)&thread_name_info);
    } __except (EXCEPTION_EXECUTE_HANDLER) {
    }
  #pragma warning(pop)
}
```

**Comment by graphitemaster** [comment]
When `Thread` and `ThreadPool` are reworked as described in #42 this will be handled. Mentioning it here so that I don't forget to.

**Comment by graphitemaster** [comment]
Fixed in 5e9f857
